### ISSUE TITLE
Improve rendering of street names

### ISSUE THREAD

**Issue by matthijsmelissen** [gravitystorm/openstreetmap-carto#497]
The following issue has been moved over from [trac](https://trac.openstreetmap.org/ticket/2991). It needs more discussion as it is not clear how this should be implemented.

Mapnik currently seems to render street names once in the middle of each piece of way, with various undesirable effects such as
- [ ] names not rendering at all if the road is split up in multiple short ways (which may happen for multiple reasons, such as different relation memberships or attributes applying just to one part of the way, such as speed limits)
- [ ] names wrapping around angles at strange places when there is enough space to render the name in a straight line
- [x] names of very long roads appearing just once, making it unclear where the boundary between two differently named sections of the same road (if any) is
- [ ] the name rendering twice if the road is made up of two separate ways which blend into each other when rendering 

This could be remedied by the following:
- Merge all pieces of road if they have the same name and highway attributes and are either joined (i.e. share a common node) or blend into each other (zoom level dependent); render the name for the entire length of the combined road
- If the resulting road is very long, repeat the names along its length multiple times:
  1) at each end and
  2) optionally in between where other roads cross, to make it clear the road continues beyond the junction 
- Where possible, pick an almost straight section (defined by minimum angle at a single point and/or minimum curve radius for consecutive points) and render the name there. Failing that, consider splitting the name string up at spaces, hyphens etc. (i.e. by the same rules that apply for line breaks) and find a straight section for each piece of the name. If that doesn't help, "massage" the straightness criteria for that road until a solution is found.

**Comment by kocio-pl** [comment]
It should be doable also without lua transformations.

Currently 28 907 out of 1 002 840 sidewalks have name:

https://taginfo.openstreetmap.org/tags/footway=sidewalk#combinations

**Comment by geostonemarten** [comment]
@kocio-pl in fact that can be uprising because this is on discussion with French other contributors delete the names because it considers that have had rendering  for the display.

**Comment by kocio-pl** [comment]
Is it also a problem for areas or only for lines? We have a `roads-area-text-name` and `paths-text-name` data layers in `project.mml` for rendering names, both include footways and could filter out these with `footway=sidewalk`.

**Comment by pnorman** [comment]
We wouldn't use Lua for this, we'd just delete the CartoCSS rules for names on footway/path/cycleway.

I'd want to see more evidence about the pros **and cons** of this before deciding if it was a good idea.

**Comment by geostonemarten** [comment]
> Is it also a problem for areas or only for lines? We have a `roads-area-text-name` and `paths-text-name` data layers in `project.mml` for rendering names, both include footways and could filter out these with `footway=sidewalk`.

I haven't check that but i can test it. 

> We wouldn't use Lua for this, we'd just delete the CartoCSS rules for names on footway/path/cycleway.
> 
> I'd want to see more evidence about the pros **and cons** of this before deciding if it was a good idea.

Ok, for me If you have a sidewalk path you have also a major road. Is it possible to check that case with overpass. Ideally It would be a good idea to use also a distance according to the zoom.

This is in relation to[Sidewalk as separate way]( https://wiki.openstreetmap.org/wiki/Sidewalks#Sidewalk_as_separate_way)

Other problematic is in relation to parallel cycleway lanes with names. I Think it's more easy for rooting if name exist on path. Anyway for the foot path (specific to blind or wheelchair) it is good practice to represent a reality with specific markings. An solution exist based on this to create specific rooting (specificly for blind with accelerated text to speech)

for exemple Versailles France zoom 17 surligned footway
![image](https://user-images.githubusercontent.com/5580128/48826160-e5f22200-ed69-11e8-9141-573d7f620683.png)
In yellow footway, in red living_street, in blue cycleway, in black highway tertiary
![image](https://user-images.githubusercontent.com/5580128/48826665-403fb280-ed6b-11e8-9a47-7495e8869f53.png)

https://www.openstreetmap.org/search?query=paris#map=17/48.80268/2.13187

I think this problem is only on zoom >= 17

**Comment by kocio-pl** [comment]
Here's what I've found in my city which I thought initially is a con, but now I think this is bad tagging - this footway is not a sidewalk (it's not on a side of the road):

https://www.openstreetmap.org/way/20931443

![screenshot_2018-11-21 openstreetmap](https://user-images.githubusercontent.com/5439713/48838657-4ba1d680-ed89-11e8-9cb0-70133824e179.png)

**Comment by SomeoneElseOSM** [comment]
> but now I think this is bad tagging 

Yes, that's doesn't look like a sidewalk - there's nothing mapped for it to be the sidewalk of.

**Comment by kocio-pl** [comment]
If we follow the logic of sidewalk being accompanying (parallel) way, which is how it's meant in general (_"a path along the side of a road"_ in Wikipedia) and in OSM definition (_"part of a highway set aside for the use of pedestrians and sometimes also cyclists, separated from the carriageway (or roadway)"_), I think it would be hard to find any cons, because the name of a main road is always expected if it exists. 

It might happen that these names are different or a main road does not have a name, but I guess this is an exception from general rule and the clutter is already visible currently.

**Comment by jragusa** [comment]
I agree to remove label of `path/footway` when associated to `footway=sidewalk`

**Comment by pnorman** [comment]
> Ok, for me If you have a sidewalk path you have also a major road. Is it possible to check that case with overpass. Ideally It would be a good idea to use also a distance according to the zoom.

This would all need to be done in CartoCSS. The only stuff that comes to mind is the buffering for identical labels, nothing elsse seems like it would relevant to what you're proposing.

> This is in relation to[Sidewalk as separate way](https://wiki.openstreetmap.org/wiki/Sidewalks#Sidewalk_as_separate_way)

We can't do anything with those relations.

**Comment by 1ec5** [comment]
>> This is in relation toSidewalk as separate way
> 
> We can't do anything with those relations.

Pretty sure that was “in relation to” as a preposition, not a reference to any kind of relation. 😉

**Comment by jeisenbe** [comment]
We could remove all names from all sidewalks, but this would lead to
problems with “sidewalks” which do not directly follow a road, or which
have a different name than the nearby road.

**Comment by kocio-pl** [comment]
I believe that would be smaller problem than a current one. Also, I'd like to check the examples to see how many of them are there and if it is that bad.

Another solution might be to make bigger spaces between labels on sidewalks, similar to latest road labels tuning - see for example https://github.com/gravitystorm/openstreetmap-carto/pull/3318#issuecomment-430470525.

**Comment by geostonemarten** [comment]
Not only big space because you need set a priority to displayed names by type of label. In addition not all footway was sidewalks. Other problems with bigger spaces is deletion of to many names

**Comment by kocio-pl** [comment]
I don't quite understand - do you think it's better to remove all the sidewalk labels than just make bigger spaces between labels?

**Comment by kocio-pl** [comment]
But we were discussing lines and areas, so I'm still confused. Could you make it more detailed?

**Comment by geostonemarten** [comment]
reread my message please. https://github.com/gravitystorm/openstreetmap-carto/issues/497#issuecomment-440572724

line and area is an other problem and is discuss in an other [issue](https://github.com/gravitystorm/openstreetmap-carto/issues/3342)

My discuss is on twice subject:

1. ignore labelling on footway with sidewalk footway=sidewalk and footway=crossing (because problem is same)
2. add buffer detection on others ways with priority classification for labelling 

In first subject we don't need labelling if this is same name as major side road because this is just a micromapping schema and this solution is better for me for many reasons. sidewalk path can uniquely exist if you have a major road in other case this is mapping error and not a display problem.

In second subject: if cycleway have same name of side highway=residential we don't need labelling cycleway.
for me priority is : motorway*>trunck*>primary*>secondary*>tertiary*>residential>=unclassified>bus_guideway>living_street>=pedestrian>road>steps>truck>path>cycleway>footway> ... others

stars for "_link"

I can't test my proposition because I haven't install server and database for that.

**Comment by geostonemarten** [comment]
@Adamant36 which convention? name can be identical on sidewalk as a separeted way. This is just an other way to create road and specific part with description. problem is on lisibility on map rendering. If you want create a map with show case on cycleway and sidewalk you reverse priority on map and the rendering.

You need separate data and rendering in fact this is done partially with @kocio-pl bigger space. But in this map you don't need name on sidewalk because there is a major road name and (micromapping) sidewalk overload redabillity
